bin/propolis-server/src/lib/migrate/destination.rs

```diff
@@ -24,105 +27,321 @@ use crate::migrate::probes;
+            info!(log, "source negotiated protocol {}", selected_protocol);
+            match super::protocol::select_protocol_from_offer(
+                &selected_protocol,
+            ) {
+                Ok(Some(selected)) => selected,
+                Ok(None) => {
+                    let offered = super::protocol::make_protocol_offer();
+                    error!(log, "source selected protocol not on offer";
+                           "offered" => &offered,
+                           "selected" => &selected_protocol);
+
+                    return Err(MigrateError::NoMatchingProtocol(
+                        selected_protocol,
+                        offered,
+                    ));
+                }
+                Err(e) => {
+                    error!(log, "source selected protocol failed to parse";
+                           "selected" => &selected_protocol);
+
+                    return Err(MigrateError::ProtocolParse(
+                        selected_protocol,
+                        e.to_string(),
+                    ));
+                }
+            }
+        }
+        x => {
+            conn.send(tungstenite::Message::Close(Some(CloseFrame {
+                code: CloseCode::Protocol,
+                reason: "did not respond to version handshake.".into(),
+            })))
+            .await?;
+            error!(
+                log,
+                "source instance failed to negotiate protocol version: {:?}", x
+            );
+            return Err(MigrateError::Initiate);
+        }
     };
 
-    if let Err(err) = proto.run().await {
-        let _ = err_tx
-            .send(MigrateTargetCommand::UpdateState(MigrationState::Error))
-            .await;
-
-        // We encountered an error, try to inform the remote before bailing
-        // Note, we don't use `?` here as this is a best effort and we don't
-        // want an error encountered during this send to shadow the run error
-        // from the caller.
-        if let Ok(e) = codec::Message::Error(err.clone()).try_into() {
-            let _ = proto.conn.send(e).await;
+    Ok(match selected {
+        Protocol::RonV0 => RonV0::new(log, migration_id, conn, local_addr),
+    })
+}
+
+/// A helper type for abstracting away type-level differences between the
+/// different phases of instance ensure.
+enum EnsureState<'a> {
+    NotStarted(VmEnsureNotStarted<'a>),
+    Active(VmEnsureActive<'a>),
+}
+
+impl<'a> EnsureState<'a> {
+    /// Moves the state machine to the `Active` state by creating VM objects,
+    /// preparing them for an incoming live migration, and activating the
+    /// created objects.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the state machine is not in the `NotStarted` state.
+    async fn activate(self) -> anyhow::Result<Self> {
+        let not_started = match self {
+            EnsureState::NotStarted(vm) => vm,
+            _ => unreachable!("can only activate a VM once"),
+        };
```

Comment:
style nit, unimportant: also write this as:
```suggestion
        let EnsureState::NotStarted(not_started) = self else {
            unreachable!("can only activate a VM once");
        };
```
which is one line shorter...but the match is also fine.

Reply:
Done in 6943610